Implement TCP reusable socket binding

canndrew opened the pull request:
This PR adds `bind_reusable` and `connect_reusable_nonblocking` methods to listeners and streams. These methods set `SO_REUSEADDR` and `SO_REUSEPORT` on the socket before binding it to the local address. `connect_reusable_nonblocking` also allows specifying the local address to make the connection from. Together these methods enable receiving incoming connections and making outgoing connections from the same port which is necessary for TCP hole punching.

The first commit here removes the `socket2`/`connect_nonblocking` features because I needed `socket2` to implement these methods and I couldn't see the point in keeping these features if `socket2` would have to be a dependency anyway. Alternatively I could scrap that commit and put the new methods behind another feature gate (a la `connect_nonblocking`) if that's preferred.

ping @cloudhead

canndrew commented:
I've fixed the rustfmt complaints. The other couple of CI failures though are nothing to do with this PR and aren't reproduced at my end.

cloudhead reviewed (commented):
Looks good to me. Can you explain why you had to change the crate features?

canndrew commented:
> Can you explain why you had to change the crate features?

The `socket2` crate was an optional dependency which was only used if the `connect_nonblocking` feature flag was enabled. Since these new methods require `socket2` I could either (a) put the new methods behind a new feature flag which also uses `socket2` or (b) make `socket2` an unconditional dependency. I chose (b) for simplicity. Without `socket2` being optional there doesn't seem to be any point in keeping the `connect_nonblocking` methods behind a feature flag, so I removed that flag too.

I can do it differently and choose (a) if that's preferable.

cloudhead commented:
Makes sense!

dr-orlovsky commented on Cargo.toml:
```diff
@@ -17,7 +17,7 @@ amplify = "4.0.0"
 cyphernet = { version = "0.3.0", features = ["ed25519", "pem", "multibase", "noise_sha2", "noise_x25519", "noise-framework", "mixnets", "dns"] }
 mio = { version = "0.8.6", optional = true }
-socket2 = { version = "0.5.2", optional = true }
+socket2 = { version = "0.5.2", features = ["all"] }
```
Since `all` can be extended with minor release, I think it is better to commit here to a specific set of features we need.

canndrew replied:
`all` is `socket2`'s only feature. I can't commit to anything more fine-grained than that.

dr-orlovsky commented on src/connection.rs:
```diff
@@ -163,6 +172,55 @@ impl NetConnection for socket2::Socket {
+            }
+            Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
+                #[cfg(feature = "log")]
+                log::error!(target: "netservices", "Can't connect to {} in a non-blocking way", remote_addr);
```
Why are we ok with skipping this error and returning socket as if it was connected?

canndrew replied:
That logic was copy-pasted from the pre-existing `connect_nonblocking` method. A successful call to `connect` on a non-blocking socket should either return zero or set errno to `EINPROGRESS` (at least according to docs) so I agree it's not correct to ignore `EWOULDBLOCK`, though maybe someone had a reason for writing it like this?

dr-orlovsky replied:
Than we can merge that and sort out this question as an independent issue. I used this probably after @cloudhead suggestion (but not sure), so I assume he had some issues with reactor/nakomoto in the past which caused that. Anyway this is the stuff which works well in current radicle, so better not to touch.

dr-orlovsky reviewed (commented):
1. I have no idea how this works, so I trust @cloudhead that the design is correct.
2. LGTM except the comments - plus not clear why we are getting rid of `socket2` forcing all users (even those who do not need this feature) to compile it. So I propose to leave the feature, make the crate optional and add `#[cfg(feature = "nonblocking")] to the all new 'reusable' APIs.

canndrew commented:
I've updated this. `socket2`/`connect_nonblocking` features are now reinstated with `connect_nonblocking` renamed to `nonblocking` since it doesn't just enable the `connect_nonblocking` method anymore.

I've also updated the `proc_macro2` version in `Cargo.lock` which should fix the codecov CI error.

dr-orlovsky reviewed (approved):
ACK a473a29126fd4784f3bb529c0774c54c289f6a53
